presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java

```diff
@@ -200,15 +218,28 @@ private JoinGraph joinWith(JoinGraph other, List<JoinNode.EquiJoinClause> joinCl
     private static class Builder
             extends PlanVisitor<Context, JoinGraph>
     {
+        // TODO When com.facebook.presto.sql.planner.optimizations.EliminateCrossJoins is removed, remove 'shallow' flag
```

Comment:
I don't see how this would work. You mean incremental construction of `MultiJoinNode`-s from `Join` and other `MultiJoinNode` with one rule? You would need to "swallow" filters and projections too (it's what supported by the old optimizer). And you would need to know when to stop and allow some second rule to decompose that into `Join`-s. It doesn't play with current `IterativeOptimizer` well, does it? Wouldn't you need two separate `IterativeOptimizer`-s that _cannot_ be combined into one?